msgcntr/messageforums-app/src/java/org/sakaiproject/tool/messageforums/ui/DiscussionMessageBean.java

```diff
@@ -449,4 +457,128 @@ public String getAuthorEid(){
 		String userEid = this.getMessage().getCreatedBy();
  		return userEid;
 	}	
+
+	private boolean isAnonymousEnabled()
+	{
+		return ServerConfigurationService.getBoolean("msgcntr.forums.anonymous.enable", true);
+	}
+
+	public void setAnonId(String anonId)
+	{
+		this.anonId = anonId;
+	}
+
+	/**
+	 * Returns the author's anonymousID regardless of whether we're in an anonymous context (unless anonymous is disabled system wide)
+	 */
+	public String getAnonId()
+	{
+		if (!isAnonymousEnabled())
+		{
+			return message.getAuthor();
+		}
+
+		if (anonId != null)
+		{
+			// anonId has already been supplied
+			return anonId;
+		}
+
+		// get / create the anonId from the AnonymousManager
+		AnonymousManager anonymousManager = (AnonymousManager)ComponentManager.get("org.sakaiproject.api.app.messageforums.AnonymousManager");
+		String siteId = ToolManager.getInstance().getCurrentPlacement().getContext();
+		anonId = anonymousManager.getOrCreateAnonId(siteId, message.getAuthorId());
+		return anonId;
+	}
+
+	public String getAnonAwareAuthor()
+	{
+		return isUseAnonymousId() ? getAnonId() : message.getAuthor();
+	}
+
+	public void setUseAnonymousId(boolean useAnonymousId)
+	{
+		this.useAnonymousId = useAnonymousId;
+	}
+
+	public boolean isUseAnonymousId()
+	{
+		if (!isAnonymousEnabled())
+		{
+			return false;
+		}
+
+		if (useAnonymousId != null)
+		{
+			// useAnonymousId has already been supplied
+			return useAnonymousId;
+		}
+
+		// Determine if the containing topic is anonymous
+		Topic topic = message.getTopic();
+		if (topic != null)
+		{
+			// Determine if topic is anonymous
+			// Topic may or may not be detached (ie. it may have been retrieved as the surrogate key of this mesage, so its only initialized attribute may be the topic ID).
+			// Retrieve the full topic only when this is an issue via a try / catch
+			Boolean postAnonymous = null;
+			try
+			{
+				postAnonymous = topic.getPostAnonymous();
+			}
+			catch (RuntimeException e)
+			{
+				// Only act on this exception if its a hibernate LazyInitializationException (without depending on Hibernate as the ORM)
+				if ("org.hibernate.LazyInitializationException".equals(e.getClass().getName()))
+				{
+					// Topic is detached, retrieve it via ForumManager
+					MessageForumsForumManager forumManager = (MessageForumsForumManager)ComponentManager.get("org.sakaiproject.api.app.messageforums.MessageForumsForumManager");
+					topic = forumManager.getTopicById(true, topic.getId());
+					postAnonymous = topic.getPostAnonymous();
+				}
+				else
+				{
+					// This is unexpected. Rethrow the exception
+					throw e;
+				}
+			}
+			if (postAnonymous)
+			{
+				// Are we supposed to reveal authors' identities to certain roles in this topic?
+				if (topic.getRevealIDsToRoles())
+				{
+					UIPermissionsManager uiPermissionsManager = (UIPermissionsManager) ComponentManager.get("org.sakaiproject.api.app.messageforums.ui.UIPermissionsManager");
```

Comment:
Maybe centralize all these ComponentManager dependency calls in AnonymousManager with methods like getUIPermissionsManager()? Easier to read and maintain instead of having them scattered around the code.